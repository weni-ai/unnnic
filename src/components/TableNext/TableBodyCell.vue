<template>
  <td class="unnnic-table-next__body-cell">
    <template v-if="getIsComponent(cell)">
      <component
        :is="cell.component"
        v-bind="cell.props"
        v-on="cell.events"
      />
    </template>
    <template v-else>
      <p
        class="unnnic-table-next__body-cell-text"
        data-testid="body-cell-text"
      >
        {{ cell }}
      </p>
    </template>
  </td>
</template>

<script>
export default {
  name: 'TableBodyCell',

  props: {
    cell: {
      type: [String, Object],
      default: '',
    },
  },
  methods: {
    getIsComponent(cell) {
      return typeof cell === 'object';
    },
  },
};
</script>

<style lang="scss" scoped>
.unnnic-table-next__body-cell-text {
  margin: 0;

  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}
</style>
