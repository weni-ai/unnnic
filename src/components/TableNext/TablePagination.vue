<template>
  <section
    class="table-pagination"
  >
    <p
      class="table-pagination__count"
    >
      {{ tablePagination }}
    </p>

    <Pagination
      :value="value"
      @input="$emit('input', $event)"
      :max="pages"
      :show="5"
    />
  </section>
</template>

<script>
import Pagination from '../Pagination/Pagination.vue';
import UnnnicI18n from '../../mixins/i18n';

export default {
  name: 'TablePagination',

  mixins: [UnnnicI18n],

  components: {
    Pagination,
  },

  props: {
    value: {
      type: Number,
      required: true,
    },
    total: {
      type: Number,
      required: false,
    },
    interval: {
      type: Number,
      required: true,
    },
  },

  data() {
    return {
      defaultTranslations: {
        pagination: {
          'pt-br': '{from} - {to} de {total}',
          en: '{from} - {to} of {total}',
          es: '{from} - {to} de {total}',
        },
      },
    };
  },

  computed: {
    tablePagination() {
      const { value, interval, total } = this;

      return this.i18n('pagination', {
        from: total === 0 ? 0 : (value - 1) * interval + 1,
        to: Math.min(value * interval, total),
        total,
      });
    },

    pages() {
      return Math.ceil(this.total / this.interval);
    },
    showTotal() {
      return typeof this.total === 'number';
    },
  },
};
</script>

<style lang="scss" scoped>
@import "../../assets/scss/unnnic.scss";

.table-pagination {
  display: flex;
  align-items: center;
  justify-content: space-between;

  padding: $unnnic-spacing-sm $unnnic-spacing-nano;

  height: min-content;

  &__count {
    color: $unnnic-color-neutral-dark;
    font-size: $unnnic-font-size-body-gt;
    font-family: $unnnic-font-family-secondary;
  }
}
</style>
