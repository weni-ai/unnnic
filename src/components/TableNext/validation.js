const validateHeader = (cell) => {
  const isContentString = typeof cell.content !== 'string';
  if (isContentString) {
    throw new Error('Each item in "headers" must have "content" as a string.');
  }

  const isSortableBoolean = 'isSortable' in cell && typeof cell.isSortable !== 'boolean';
  if (isSortableBoolean) {
    throw new Error('Each item in "headers" that contains “isSortable” must assign it as boolean');
  }
};

export const validateHeaders = (headers) => {
  if (!Array.isArray(headers)) {
    throw new Error('Property headers needs to be an array.');
  }

  if (headers.length === 0) {
    throw new Error('Property headers must not to be an empty array.');
  }

  headers.forEach(validateHeader);

  return true;
};

const validateRow = (cell) => {
  const isContentArray = !('content' in cell) || !Array.isArray(cell.content);
  if (isContentArray) {
    throw new Error('Each item in "rows" must have "content" as an array.');
  }

  const isLinkObject = 'link' in cell && typeof cell.link !== 'object' && typeof cell.link !== 'undefined';
  if (isLinkObject) {
    throw new Error('Each item in "rows" that contains “link” must assign it as object');
  }

  const isLinkWithoutUrl = cell.link && !('url' in cell.link);
  if (isLinkWithoutUrl) {
    throw new Error('Each item in "rows" with "link" must contain "url".');
  }
};

export const validateRows = (rows) => {
  if (!Array.isArray(rows)) {
    throw new Error('Property rows needs to be an array.');
  }

  rows.forEach(validateRow);

  return true;
};
