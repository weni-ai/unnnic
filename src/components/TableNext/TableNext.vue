<template>
  <table class="unnnic-table-next">
    <thead class="unnnic-table-next__header">
      <tr class="unnnic-table-next__header-row">
        <th
          v-for="(cell, index) of headers"
          :key="cell.content + index"
          class="unnnic-table-next__header-cell"
        >
          {{ cell.content }}
        </th>
      </tr>
    </thead>

    <tbody class="unnnic-table-next__body">
      <tr v-if="isLoading" class="unnnic-table-next__body-row">
        <img
          class="unnnic-table-next__body-cell--loading"
          src="../../assets/icons/weni-loading.svg"
          height="40"
        />
      </tr>
      <template v-else-if="rows.length">
        <tr
          v-for="(row, index) of rows"
          :key="row.content + index"
          class="unnnic-table-next__body-row"
        >
          <a
            v-if="row.link"
            class="unnnic-table-next__body-row--redirect"
            :href="row.link.url"
            :target="row.link.target || '_blank'"
          >
            <TableBodyCell
              v-for="(cell, index) of row.content"
              :key="cell + index"
              class="unnnic-table-next__body-cell"
              :cell="cell"
            />
          </a>
          <template v-else>
            <TableBodyCell
              v-for="(cell, index) of row.content"
              :key="cell + index"
              class="unnnic-table-next__body-cell"
              :cell="cell"
            />
          </template>
        </tr>
      </template>
      <tr v-else class="unnnic-table-next__body-row">
        <td class="unnnic-table-next__body-cell">
          <p class="unnnic-table-next__body-cell-text">
            {{ i18n("without_results") }}
          </p>
        </td>
      </tr>
    </tbody>
    <TablePagination
      :value="pagination"
      @input="$emit('update:pagination', $event)"
      :total="treatedPaginationTotal"
      :interval="rows.length"
    />
  </table>
</template>

<script>
import { validateHeaders, validateRows } from './validation';
import TableBodyCell from './TableBodyCell.vue';
import TablePagination from './TablePagination.vue';
import UnnnicI18n from '../../mixins/i18n';

export default {
  name: 'UnnnicTableNext',

  mixins: [UnnnicI18n],

  components: {
    TableBodyCell,
    TablePagination,
  },

  props: {
    /**
     * @typedef {Array} HeaderItem
     * @property {string} content - The content of the header cell.
     * @property {boolean|undefined} isSortable - Indicates if the cell is enabled for sorting.
     */

    /**
     * Array of header items.
     * @type {HeaderItem[]}
     */
    headers: {
      type: Array,
      default: () => [],
      validator: validateHeaders,
    },

    /**
     * @typedef {Array} RowItem
     * @property {Array} content - The content of the row cell.
     * @property {object|undefined} link - Link object to redirect the row cell.
     */

    /**
     * Array of row items.
     * @type {RowItem[]}
     */
    rows: {
      type: Array,
      default: () => [],
      validator: validateRows,
    },

    pagination: {
      type: Number,
      default: 1,
    },
    paginationTotal: {
      type: Number,
      default: 1,
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      defaultTranslations: {
        without_results: {
          'pt-br': 'Nenhum resultado correspondente',
          en: 'No matching results',
          es: 'No hay resultados coincidentes',
        },
      },
    };
  },

  computed: {
    treatedPaginationTotal() {
      return this.rows.length === 0 ? 0 : this.paginationTotal;
    },
  },
};
</script>

<style lang="scss" scoped>
@import "../../assets/scss/unnnic.scss";

$tableBorder: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

.unnnic-table-next {
  border-spacing: 0;

  overflow: hidden;

  width: 100%;

  &__header {
    &-row {
      @extend %base-row;
    }

    &-cell {
      @extend %base-cell;

      padding: $unnnic-spacing-ant $unnnic-spacing-sm;

      height: 100%;

      box-sizing: border-box;
      border: $tableBorder;
      background-color: $unnnic-color-neutral-light;

      font-weight: $unnnic-font-weight-bold;

      &:first-of-type {
        border-radius: $unnnic-border-radius-sm 0 0 0;
      }

      &:not(:first-of-type) {
        border-left: none;
      }

      &:last-of-type {
        border-radius: 0 $unnnic-border-radius-sm 0 0;
      }
    }
  }

  &__body {
    &-row {
      @extend %base-row;

      border: $tableBorder;
      border-collapse: collapse;
      border-top: none;

      &--redirect {
        @extend %base-row;
        text-decoration: none;

        &:hover {
          background-color: $unnnic-color-neutral-lightest;
        }

        :visited {
          color: $unnnic-color-neutral-dark;
        }
      }

      &:last-of-type {
        border-radius: 0 0 $unnnic-border-radius-sm $unnnic-border-radius-sm;
      }
    }

    &-cell {
      @extend %base-cell;

      &-text {
        margin: 0;

        overflow: hidden;
        white-space: nowrap;
        text-overflow: ellipsis;

        padding: $unnnic-spacing-sm;
      }
    }
    &-cell--loading {
      margin: $unnnic-spacing-xl 0;
      padding: 0;

      width: 100%;

      pointer-events: none;
    }
  }

  %base-cell {
    border-collapse: collapse;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    line-height: $unnnic-line-height-small * 5.5;
    text-align: left;
    color: $unnnic-color-neutral-dark;

    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
  }

  %base-row {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(0, 1fr));
    align-items: center;
  }
}
</style>
