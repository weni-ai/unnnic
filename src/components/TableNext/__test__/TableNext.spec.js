import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from 'vitest';
import TableNext from '../TableNext.vue';
import TablePagination from '../TablePagination.vue';

describe('TableNext.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(TableNext, {
      props: { headers, rows },
    });
  });

  const headers = [{ content: 'Header 1' }, { content: 'Header 2', size: 2 }];
  const rows = [
    { content: ['Row 1 Col 1', 'Row 1 Col 2'] },
    {
      content: ['Row 2 Col 1', 'Row 2 Col 2'],
      link: { url: '#' },
    },
    {
      content: ['Row 3 Col 1', 'Row 3 Col 2'],
      link: { url: '#', target: '_self' },
    },
  ];

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('Header rendering', () => {
    it('renders headers correctly', () => {
      const headerCells = wrapper.findAll('[data-testid="header-cell"]');
      expect(headerCells).toHaveLength(headers.length);
      headers.forEach((header, index) => {
        expect(headerCells[index].text()).toBe(header.content);
      });
    });

    it('renders headers with correct size', () => {
      const headerRow = wrapper.find('[data-testid="header-row"]');
      const gridTemplateColumns = headerRow.attributes('style');
      expect(gridTemplateColumns).toContain('1fr 2fr');
      // The first column uses the default size of 1fr,
      // and the second column is set to 2fr as specified in the header configuration.
    });
  });

  describe('Row rendering and behaviors', () => {
    it('renders rows correctly', () => {
      const bodyRows = wrapper.findAll('[data-testid="body-row"]');
      expect(bodyRows).toHaveLength(rows.length);

      rows.forEach((row, rowIndex) => {
        const bodyCells = bodyRows[rowIndex].findAll(
          '[data-testid="body-cell"]',
        );
        expect(bodyCells).toHaveLength(row.content.length);

        row.content.forEach((cell, cellIndex) => {
          expect(bodyCells[cellIndex].text()).toBe(cell);
        });
      });
    });

    it('renders rows with links correctly', () => {
      const linkedRows = wrapper.findAll('[data-testid="body-row-redirect"]');
      expect(linkedRows).toHaveLength(2);

      const firstLink = linkedRows[0];
      expect(firstLink.attributes('href')).toBe('#');
      expect(firstLink.attributes('target')).toBe('_blank');

      const secondLink = linkedRows[1];
      expect(secondLink.attributes('href')).toBe('#');
      expect(secondLink.attributes('target')).toBe('_self');
    });

    it('renders "no matching results" message when there are no rows', async () => {
      await wrapper.setProps({ headers, rows: [] });

      const noResultsMessage = wrapper.find('[data-testid="body-cell-text"]');
      expect(noResultsMessage.text()).toBe('No matching results');
    });
  });

  describe('Loading state', () => {
    it('renders loading state', async () => {
      await wrapper.setProps({ headers, isLoading: true });
      const loadingImage = wrapper.find('[data-testid="body-row-loading"]');
      expect(loadingImage.exists()).toBe(true);
    });
  });

  describe('Pagination', () => {
    it('emits update:pagination event on pagination change', async () => {
      const paginationComponent = wrapper.findComponent(TablePagination);
      await paginationComponent.vm.$emit('update:model-value', 2);

      expect(wrapper.emitted()['update:pagination'][0]).toEqual([2]);
    });
  });
});
