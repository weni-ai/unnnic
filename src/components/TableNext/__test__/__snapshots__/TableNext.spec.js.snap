// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableNext.vue > matches the snapshot 1`] = `
"<table data-v-2143c794="" class="unnnic-table-next">
  <thead data-v-2143c794="" class="unnnic-table-next__header" data-testid="header">
    <tr data-v-2143c794="" class="unnnic-table-next__header-row" data-testid="header-row" style="grid-template-columns: 1fr 2fr;">
      <th data-v-2143c794="" class="unnnic-table-next__header-cell" data-testid="header-cell">Header 1
        <!--v-if-->
      </th>
      <th data-v-2143c794="" class="unnnic-table-next__header-cell" data-testid="header-cell">Header 2
        <!--v-if-->
      </th>
    </tr>
  </thead>
  <tbody data-v-2143c794="" class="unnnic-table-next__body" data-testid="body">
    <tr data-v-2143c794="" class="unnnic-table-next__body-row" data-testid="body-row" style="grid-template-columns: 1fr 2fr;">
      <td data-v-fd58013b="" data-v-2143c794="" class="unnnic-table-next__body-cell unnnic-table-next__body-cell unnnic-table-next__body-cell--sm" data-testid="body-cell">
        <p data-v-fd58013b="" class="unnnic-table-next__body-cell-text" data-testid="body-cell-text">Row 1 Col 1</p>
      </td>
      <td data-v-fd58013b="" data-v-2143c794="" class="unnnic-table-next__body-cell unnnic-table-next__body-cell unnnic-table-next__body-cell--sm" data-testid="body-cell">
        <p data-v-fd58013b="" class="unnnic-table-next__body-cell-text" data-testid="body-cell-text">Row 1 Col 2</p>
      </td>
    </tr>
    <tr data-v-2143c794="" class="unnnic-table-next__body-row" data-testid="body-row" style="grid-template-columns: auto;"><a data-v-2143c794="" class="unnnic-table-next__body-row--redirect" data-testid="body-row-redirect" style="grid-template-columns: 1fr 2fr;" href="#" target="_blank">
        <td data-v-fd58013b="" data-v-2143c794="" class="unnnic-table-next__body-cell unnnic-table-next__body-cell unnnic-table-next__body-cell--sm" data-testid="body-cell">
          <p data-v-fd58013b="" class="unnnic-table-next__body-cell-text" data-testid="body-cell-text">Row 2 Col 1</p>
        </td>
        <td data-v-fd58013b="" data-v-2143c794="" class="unnnic-table-next__body-cell unnnic-table-next__body-cell unnnic-table-next__body-cell--sm" data-testid="body-cell">
          <p data-v-fd58013b="" class="unnnic-table-next__body-cell-text" data-testid="body-cell-text">Row 2 Col 2</p>
        </td>
      </a></tr>
    <tr data-v-2143c794="" class="unnnic-table-next__body-row" data-testid="body-row" style="grid-template-columns: auto;"><a data-v-2143c794="" class="unnnic-table-next__body-row--redirect" data-testid="body-row-redirect" style="grid-template-columns: 1fr 2fr;" href="#" target="_self">
        <td data-v-fd58013b="" data-v-2143c794="" class="unnnic-table-next__body-cell unnnic-table-next__body-cell unnnic-table-next__body-cell--sm" data-testid="body-cell">
          <p data-v-fd58013b="" class="unnnic-table-next__body-cell-text" data-testid="body-cell-text">Row 3 Col 1</p>
        </td>
        <td data-v-fd58013b="" data-v-2143c794="" class="unnnic-table-next__body-cell unnnic-table-next__body-cell unnnic-table-next__body-cell--sm" data-testid="body-cell">
          <p data-v-fd58013b="" class="unnnic-table-next__body-cell-text" data-testid="body-cell-text">Row 3 Col 2</p>
        </td>
      </a></tr>
  </tbody>
  <section data-v-e29458bb="" data-v-2143c794="" class="table-pagination">
    <p data-v-e29458bb="" class="table-pagination__count" data-testid="count">1 - 1 of 1</p>
    <div data-v-0eeff8cb="" data-v-e29458bb="" class="pagination"><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="previous-button" disabled="" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary unnnic-button--icon-on-center">
        <!--v-if-->
        <!--v-if--><svg data-v-26446d8e="" data-v-ceff2a60="" width="40" height="40" viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg" class="unnnic-icon unnnic-icon__size--ant unnnic-icon-scheme--neutral-clean" data-testid="icon-center" style="visibility: visible;">
          <path d="M26.8368 35.0005C26.5868 35.0005 26.3506 34.903 26.1743 34.7255L12.7743 21.3255C12.4206 20.9717 12.2256 20.5017 12.2256 20.0017C12.2256 19.5017 12.4193 19.0305 12.7731 18.6767L26.1743 5.27549C26.3506 5.09799 26.5868 5.00049 26.8368 5.00049C27.0868 5.00049 27.3231 5.09799 27.4993 5.27549C27.6768 5.45174 27.7743 5.68799 27.7743 5.93799C27.7743 6.18799 27.6768 6.42424 27.4993 6.60049L14.0993 20.0005L27.4993 33.4005C27.6768 33.578 27.7743 33.813 27.7743 34.063C27.7743 34.313 27.6768 34.5492 27.4993 34.7255C27.3218 34.9017 27.0868 35.0005 26.8368 35.0005Z" fill="#3B414D" class="primary"></path>
        </svg><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> </span>
        <!--v-if-->
      </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="page-button" class="unnnic-button unnnic-button--size-small unnnic-button--secondary">
        <!--v-if-->
        <!--v-if-->
        <!--v-if--><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> 1</span>
        <!--v-if-->
      </button><button data-v-ceff2a60="" data-v-0eeff8cb="" data-test="next-button" disabled="" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary unnnic-button--icon-on-center">
        <!--v-if-->
        <!--v-if--><svg data-v-26446d8e="" data-v-ceff2a60="" width="40" height="40" viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg" class="unnnic-icon unnnic-icon__size--ant unnnic-icon-scheme--neutral-clean" data-testid="icon-center" style="visibility: visible;">
          <path d="M13.1635 35.0005C12.9135 35.0005 12.6772 34.903 12.501 34.7255C12.136 34.3605 12.136 33.7655 12.501 33.3992L25.901 20.0005L12.501 6.60049C12.3235 6.42424 12.226 6.18799 12.226 5.93799C12.226 5.68799 12.3235 5.45174 12.501 5.27549C12.6772 5.09799 12.9135 5.00049 13.1635 5.00049C13.4135 5.00049 13.6497 5.09799 13.826 5.27549L27.226 18.6755C27.956 19.4055 27.9572 20.5942 27.2272 21.3242L13.826 34.7255C13.6497 34.903 13.4135 35.0005 13.1635 35.0005Z" fill="#3B414D" class="primary"></path>
        </svg><span data-v-ceff2a60="" class="unnnic-button__label" style="visibility: visible;" data-testid="button-label"> </span>
        <!--v-if-->
      </button></div>
  </section>
</table>"
`;
